include/envoy/filesystem/filesystem.h

```diff
@@ -61,5 +61,30 @@ class Watcher {
 
 typedef std::unique_ptr<Watcher> WatcherPtr;
 
+/**
+ * Abstraction for listing a directory.
+ * TODO(sesmith177): replace with std::filesystem::directory_iterator once we move to C++17
```

Comment:
Would it make sense to have `DirectoryIterator` implement a C++ iterator interface for now?